src/modules/formBuilder/components/formBuilderUtil.ts

```diff
@@ -90,6 +90,137 @@ export const updateFormItem = (
   return copy;
 };
 
+function moveIntoGroup(
+  itemToMove: FormItem,
+  group: FormItem,
+  addMethod: 'push' | 'unshift',
+  thisLayer: FormItem[],
+  currentIndex: number
+) {
+  if (!group.item) group.item = [];
+  group.item[addMethod](itemToMove);
+  thisLayer.splice(currentIndex, 1);
+  return true;
+}
+
+function swapItems(index1: number, index2: number, thisLayer: FormItem[]) {
+  [thisLayer[index1], thisLayer[index2]] = [
+    thisLayer[index2],
+    thisLayer[index1],
+  ];
+  return true;
+}
+
+function moveToParent(
+  itemToMove: FormItem,
+  parentLayer: FormItem[],
+  newIndex: number,
+  thisLayer: FormItem[],
+  currentIndex: number
+) {
+  parentLayer.splice(newIndex, 0, itemToMove);
+  thisLayer.splice(currentIndex, 1);
+  return true;
+}
+
+export const reorderFormItems = (
+  formDefinition: FormDefinitionJson,
+  itemToMove: FormItem,
+  direction: 'up' | 'down'
+) => {
+  const copy = cloneDeep(formDefinition);
+  const toExpand: string[] = [];
+
+  function recurseFindAndMove(
```

Comment:
Overall I think the code I've added in this file is pretty messy and I kept wondering if there were a more graceful way to do it. We could think of this conceptually as a depth-first search to unwrap the whole tree into a list, swapping two consecutive items, and then re-wrapping it into a tree... but I wasn't convinced that taking that approach would actually be performant, so I ended up with this messy morass of special case if-statements. I tried to add lots of comments for clarity.

Reply:
Hmm, I wonder if it would be more graceful with React Hook Forms. I'm looking at  [useFieldArray](https://react-hook-form.com/docs/usefieldarray) (and maybe [setValue](https://react-hook-form.com/docs/useform/setvalue)).

RHF syntax would have the following style of "name"s for each item
```
item.0
item.0.item.0
item.0.item.1.item.0
item.0.item.1.item.1
item.0.item.2
item.1
item.2
item.2.item.0
```


useFieldArray provides functions to prepend/append/remove/swap within a given array. I used useFieldArray to add/remove items from the EnableWhen condition array in https://github.com/greenriver/hmis-frontend/pull/788..

In this context, each Group Item would have it's own FieldArray
```
  // In the context of the component that is rendering the Group that is at location `item.0.item.1`:
  const { prepend, append, remove, swap } = useFieldArray({ control, name: "item.0.item.1.item" });

  // can easily use "swap" to in onClick handlers to move items up and down within this array

  // moving the last item "down" or first item "up" feels trickier
   const movefirstItemUp = (item) => {
     insertAboveGroup(item) // insert above this group. this prop would need to be passed into this component from the parent group, I think. This would encapsulate logic for which nesting level it should land at (?) based on whether it's a Group or a question
     remove(item); // remove from this group
   }
   const moveLastItemDown = (item) => {
     // same but other direction
   }
```


This is a super rough idea, but if it works it could cut down on code complexity. Maybe we can take a stab at it in a pairing session?

Reply:
Yeah this looks interesting. I didn't even think of using rhf. I would be glad to take a stab at it together! I want to take a closer look at your PR #788 first, maybe it will help me get a better sense of `useFieldArray`.